mod/state-transition/pkg/core/state_processor_staking.go

```diff
@@ -204,6 +206,118 @@ func (sp *StateProcessor[
 		math.Gwei(sp.cs.MaxEffectiveBalance()),
 	)
 
+	// BeaconKit enforces a cap on the validator set size. If the deposit
+	// breaches the cap, we find the validator with the smallest stake and
+	// mark it as withdrawable so that it will be eventually evicted and
+	// its deposits returned.
+	capHit, err := sp.isValidatorCapHit(st)
+	if err != nil {
+		return err
+	}
+	if !capHit {
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	// Adding the validator would breach the cap. Find the validator
+	// with the smallest stake among current and candidate validators
+	// and kit it out.
+	var currSmallestVal ValidatorT
+	currSmallestVal, err = sp.findSmallestValidator(st)
+	if err != nil {
+		return err
+	}
+
+	var slot math.Slot
+	slot, err = st.GetSlot()
+	if err != nil {
+		return err
+	}
+	epoch := sp.cs.SlotToEpoch(slot)
+
+	if val.GetEffectiveBalance() <= currSmallestVal.GetEffectiveBalance() {
+		// in case of tie-break among candidate validator we prefer
+		// existing one so we mark candidate as withdrawable
+		val.SetWithdrawableEpoch(epoch)
+		return sp.addValidatorInternal(st, val, dep.GetAmount())
+	}
+
+	// mark exiting validator for eviction and add candidate
+	currSmallestVal.SetWithdrawableEpoch(epoch)
```

Comment:
question: should this be `epoch+1`? I.e. should we keep the validator for the current epoch and evict it at the start of the next? Or should we evict it immediately?